doc/reference/trd-storage-permissions.md

```diff
@@ -0,0 +1,183 @@
+Application Persistent Data Storage Permissions
+===============================================
+
+**TRD:** <br/>
+**Working Group:** Kernel<br/>
+**Type:** Documentary<br/>
+**Status:** Draft <br/>
+**Author:** Brad Campbell<br/>
+**Draft-Created:** 2024/06/06<br/>
+**Draft-Modified:** 2024/06/06<br/>
+**Draft-Version:** 1<br/>
+**Draft-Discuss:** devel@lists.tockos.org<br/>
+
+
+Abstract
+-------------------------------
+
+Tock supports storing persistent state for applications, and all persistent
+state in Tock is identified based on the application that stored it. Tock
+supports permissions for persistent state, allowing for the kernel to restrict
+which applications can store state and which applications can read stored state.
+This TRD describes the permissions architecture for persistent state in Tock.
+This document is in full compliance with [TRD1][TRD1].
+
+
+1 Introduction
+-------------------------------
+
+Tock applications need to be able to store persistent state. Additionally,
+applications need to be able to keep data private from other applications. The
+kernel should also be able to allow specific applications to read and modify
+state from other applications.
+
+This requires a method for assigning applications persistent identifiers,
+a mechanism for granting storage permissions to specific applications,
+and kernel abstractions for implementing storage capsules that respect
+the storage permissions.
+
+
+2 Scope
+-------------------------------
+
+This document only describes the permission architecture in Tock for supporting
+application persistent storage. This document does not include specific types of
+persistent storage (e.g., flash, FRAM, etc.) or storage access abstractions
+(e.g., block-access, byte-access, etc.).
+
+
+3 Stored State Identifiers
+-------------------------------
+
+All shared persistent storage implementations must store a 32 bit identifier
+with each stored object to mark the application that created the stored object.
+
+When applications write data, their ShortId must be used as the identifier. When
+the kernel writes data, the identifier must be 0.
+
+
+4 Permissions
+-------------------------------
+
+All persistent application data is labeled based on the application which wrote
+the data.
+Applications can read and modify data with suitable
+permissions.
+
+There are three types of permissions:
+
+1. **Write**: The application can write data.
+1. **Read**: The application can read data.
+1. **Modify**: The application can modify existing data.
```

Comment:
Would `Write` be clearer as `Append` given how these permissions break down?

```suggestion
1. **Append**: The application can write new data.
1. **Read**: The application can read data.
1. **Modify**: The application can modify existing data.
```

Reply:
Append to what?

Reply:
The storage region?

IIUC, `write` as defined here will always consume new bytes of storage space. Only `modify` can write already used space? That feels more like an append operation?

Reply:
I think we have to keep write. It's just the obvious counterpart to read.

We could remove modify. That would make it a little more difficult to express the case where an app can overwrite but not write new, but it wouldn't be impossible.

Reply:
> I think we have to keep write. It's just the obvious counterpart to read.

And yet if I have read and write permissions I can't write data that I can read? If there's no data there already, I can't write at all?

I'm going to disagree with "just obvious" as the semantics of write here are not just obvious at all. (Indeed, what existing non-Tock system does this definition of "write" match?)

Reply:
> And yet if I have read and write permissions I can't write data that I can read?

If parsed as "there exists data which I can read, but I can't overwrite/change/modify that data", correct. Just like if we both have accounts on the same unix system I might be able to read your files but not change them.

If parsed as "can I always read data that I wrote": no. You might have write permission, but not read permission to your own stored state. This supports the logging case where the logger is write-only. (Sidenote: I personally think applications should always be able to read their own state, but, I think it's ok if I just use storage permissions implementations which allow this and not codify that here.)

> If there's no data there already, I can't write at all?

If you do not have the write permission, but do have modify, then yes. Maybe some app cannot allocate new space for itself, but can update state where storage has already been allocated.


> I'm going to disagree with "just obvious" as the semantics of write here are not just obvious at all. (Indeed, what existing non-Tock system does this definition of "write" match?)

I'm not necessarily saying that the semantics are obvious. I am saying that I don't want to have to teach this to people and have to explain that "oh yeah, Tock doesn't have any storage permission called 'write'; we do have 'read' and 'modify', now let me explain how applications can write data".